rust/chains/hyperlane-ethereum/src/merkle_tree_hook.rs

```diff
@@ -0,0 +1,282 @@
+        // ``SequenceIndexer` and `Indexer`.
+        let tip = self.get_finalized_block_number().await?;
+        Ok((None, tip))
+    }
+}
+
+/// A reference to a Mailbox contract on some Ethereum chain
+#[derive(Debug)]
+pub struct EthereumMerkleTreeHook<M>
+where
+    M: Middleware,
+{
+    contract: Arc<MerkleTreeHookContract<M>>,
+    domain: HyperlaneDomain,
+    provider: Arc<M>,
+}
+
+impl<M> EthereumMerkleTreeHook<M>
+where
+    M: Middleware,
+{
+    /// Create a reference to a mailbox at a specific Ethereum address on some
+    /// chain
+    pub fn new(provider: Arc<M>, locator: &ContractLocator) -> Self {
+        Self {
+            contract: Arc::new(MerkleTreeHookContract::new(
+                locator.address,
+                provider.clone(),
+            )),
+            domain: locator.domain.clone(),
+            provider,
+        }
+    }
+}
+
+impl<M> HyperlaneChain for EthereumMerkleTreeHook<M>
+where
+    M: Middleware + 'static,
+{
+    fn domain(&self) -> &HyperlaneDomain {
+        &self.domain
+    }
+
+    fn provider(&self) -> Box<dyn HyperlaneProvider> {
+        Box::new(EthereumProvider::new(
+            self.provider.clone(),
+            self.domain.clone(),
+        ))
+    }
+}
+
+impl<M> HyperlaneContract for EthereumMerkleTreeHook<M>
+where
+    M: Middleware + 'static,
+{
+    fn address(&self) -> H256 {
+        self.contract.address().into()
+    }
+}
+
+#[async_trait]
+impl<M> MerkleTreeHook for EthereumMerkleTreeHook<M>
+where
+    M: Middleware + 'static,
+{
+    #[instrument(skip(self))]
+    async fn latest_checkpoint(&self, maybe_lag: Option<NonZeroU64>) -> ChainResult<Checkpoint> {
+        let lag = maybe_lag.map(|v| v.get()).unwrap_or(0).into();
+
+        let fixed_block_number: BlockNumber = self
+            .provider
+            .get_block_number()
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .saturating_sub(lag)
+            .into();
+
+        let (root, index) = self
+            .contract
+            .latest_checkpoint()
+            .block(fixed_block_number)
+            .call()
+            .await?;
+        Ok(Checkpoint {
+            merkle_tree_hook_address: self.address(),
+            mailbox_domain: self.domain.id(),
+            root: root.into(),
+            index,
+        })
+    }
+
+    #[instrument(skip(self))]
+    #[allow(clippy::needless_range_loop)]
+    async fn tree(&self, maybe_lag: Option<NonZeroU64>) -> ChainResult<IncrementalMerkle> {
+        let lag = maybe_lag.map(|v| v.get()).unwrap_or(0).into();
+
+        let fixed_block_number: BlockNumber = self
+            .provider
+            .get_block_number()
+            .await
+            .map_err(ChainCommunicationError::from_other)?
+            .saturating_sub(lag)
+            .into();
+
+        // TODO: implement From<Tree> for IncrementalMerkle
```

Comment:
may not wanna do this because the `Tree` type is ethereum-specific and we should avoid leaking types like that into vm-agnostic code

Reply:
isnt this the ethereum crate?

Reply:
I was originally saying this bc `IncrementalMerkle` is in hyperlane-core and this would require adding `hyperlane-ethereum` as a dependency to `hyperlane-core` (which imo we can't do), but maybe we could instead just do `Into<IncrementalMerkle> for Tree` inside ethers-ethereum

Reply:
implemented inside the merkle tree hook module (this same file) since that's where it's used

Reply:
btw that comment was already there. in general I kept the merkle tree code intact but it's good we get to revisit these